kcp/controllers/watcher_controller.go

```diff
@@ -74,3 +114,312 @@ func (r *WatcherReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	namespace := obj.GetNamespace()
+	vsName := obj.GetName()
+	listenerVirtualService, apiErr := istioClientSet.NetworkingV1beta1().
+		VirtualServices(namespace).Get(ctx, vsName, metav1.GetOptions{})
+	ready, err := util.IstioResourcesErrorCheck(istioGatewayGVR, apiErr)
+	if !ready {
+		return err
+	}
+	if errors.IsNotFound(apiErr) {
+		//create virtual service with config from CR
+		vs := &istioclientapiv1beta1.VirtualService{}
+		vs.SetName(vsName)
+		vs.SetNamespace(namespace)
+		util.UpdateVirtualServiceConfig(vs, obj, istioGatewayResourceName)
+		_, err := istioClientSet.NetworkingV1beta1().
+			VirtualServices(namespace).Create(ctx, vs, metav1.CreateOptions{})
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+	//check if config already exists
+	if util.IsVirtualServiceConfigChanged(listenerVirtualService, obj, istioGatewayResourceName) {
+		util.UpdateVirtualServiceConfig(listenerVirtualService, obj, istioGatewayResourceName)
+		_, err = istioClientSet.NetworkingV1beta1().
+			VirtualServices(namespace).Update(ctx, listenerVirtualService, metav1.UpdateOptions{})
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *WatcherReconciler) createConfigMapForCR(ctx context.Context, obj *componentv1alpha1.Watcher) error {
+	watcherCRLabels := obj.GetLabels()
+	_, ok := watcherCRLabels[defaultOperatorWatcherCRLabel]
+	if ok {
+		//watcher CR belongs to a KCP operator which applies on all Kymas.
+		//So no need to create configMap for it!
+		return nil
+	}
+	//create empty config map for CR
+	watcherObjKey := client.ObjectKeyFromObject(obj)
+	configMap := &v1.ConfigMap{}
+	err := r.Get(ctx, watcherObjKey, configMap)
+	if err != nil && !errors.IsNotFound(err) {
+		return fmt.Errorf("failed to send get config map request to API server: %w", err)
+	}
+	if errors.IsNotFound(err) {
+		configMap.SetName(watcherObjKey.Name)
+		configMap.SetNamespace(watcherObjKey.Namespace)
+		err = r.Create(ctx, configMap)
+		if err != nil {
+			return fmt.Errorf("failed to send create config map request to API server: %w", err)
+		}
+	}
+	return nil
+}
+
+func (r *WatcherReconciler) updateWatcherCRStatus(ctx context.Context, obj *componentv1alpha1.Watcher,
+	state componentv1alpha1.WatcherState, msg string) error {
+	obj.Status.State = state
+	switch state { //nolint:exhaustive
+	case componentv1alpha1.WatcherStateReady:
+		util.AddReadyCondition(obj, componentv1alpha1.ConditionStatusTrue, msg)
+	case "":
+		util.AddReadyCondition(obj, componentv1alpha1.ConditionStatusUnknown, msg)
+	default:
+		util.AddReadyCondition(obj, componentv1alpha1.ConditionStatusFalse, msg)
+	}
+	return r.Status().Update(ctx, obj.SetObservedGeneration())
+}
+
+func (r *WatcherReconciler) HandleDeletingState(ctx context.Context, logger logr.Logger,
+	obj *componentv1alpha1.Watcher) error {
+	err := r.deleteConfigMapForCR(ctx, obj)
+	if err != nil {
+		return r.updateWatcherCRErrStatus(ctx, logger, err, obj, "failed to create config map")
```

Comment:
Error message seems to be wrong. I would suggest `failed to delete configmap`
Same goes for the next upcoming lines